### PR TITLE
Fixes configured default shell not used when connecting to remote

### LINKED ISSUE #175107: Configured default shell not used when reconnecting to Codespace
Does this issue occur when all extensions are disabled?: Yes

Version: 1.75.1 (Universal)
Commit: 441438abd1ac652551dbe4d408dfcec8a499b8bf
Date: 2023-02-08T21:34:59.000Z
Electron: 19.1.9
Chromium: 102.0.5005.194
Node.js: 16.14.2
V8: 10.2.154.23-electron.0
OS: Darwin arm64 22.3.0
Sandboxed: Yes

When reconnecting to an existing Codespace, my `"terminal.integrated.defaultProfile.linux"` configuration is ignored. Instead of `zsh`, the terminal always starts `bash`.

The same/a similar issue used to exist in early 2022 but was presumably fixed with https://github.com/microsoft/vscode/pull/154290, however this appears to have regressed over the last weeks/months.

Steps to Reproduce:

1. Configure `"terminal.integrated.defaultProfile.linux": "zsh"`
2. Create a Codespace & stop it
3. Reconnect to the Codespace
4. See that the terminal opens using `bash`

### PR DESCRIPTION
**Pull request by jeanp413**
Fixes #175107

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminalInstanceService.ts [hunk 1]
```diff
@@ -17,27 +17,37 @@ import { Emitter, Event } from 'vs/base/common/event';
...
 		this._terminalShellTypeContextKey = TerminalContextKeys.shellType.bindTo(this._contextKeyService);
 		this._terminalInRunCommandPicker = TerminalContextKeys.inTerminalRunCommandPicker.bindTo(this._contextKeyService);
 		this._terminalSuggestWidgetVisibleContextKey = TerminalContextKeys.suggestWidgetVisible.bindTo(this._contextKeyService);
 		this._configHelper = _instantiationService.createInstance(TerminalConfigHelper);
+
+
+		for (const remoteAuthority of [undefined, _environmentService.remoteAuthority]) {
+			let resolve: () => void;
+			const p = new Promise<void>(r => resolve = r);
+			this._backendRegistration.set(remoteAuthority, { promise: Promise.race([p, this._lifecycleService.when(LifecyclePhase.Eventually)]), resolve: resolve! });
```
Comment (jeanp413) [inline src/vs/workbench/contrib/terminal/browser/terminalInstanceService.ts:49]: `RemoteTerminalBackendContribution` is registered to be created on `restored` phase but instantiation only occurs when idle, so we need to set an upper limit in this case `eventually` phase

File: src/vs/workbench/contrib/terminal/browser/terminalInstanceService.ts [hunk 2]
```diff
@@ -16,28 +16,36 @@ import { URI } from 'vs/base/common/uri';
...
 		this._terminalInRunCommandPicker = TerminalContextKeys.inTerminalRunCommandPicker.bindTo(this._contextKeyService);
 		this._terminalSuggestWidgetVisibleContextKey = TerminalContextKeys.suggestWidgetVisible.bindTo(this._contextKeyService);
 		this._configHelper = _instantiationService.createInstance(TerminalConfigHelper);
+
+
+		for (const remoteAuthority of [undefined, _environmentService.remoteAuthority]) {
+			let resolve: () => void;
+			const p = new Promise<void>(r => resolve = r);
+			this._backendRegistration.set(remoteAuthority, { promise: p, resolve: resolve! });
+		}
```
Comment (jeanp413) [inline src/vs/workbench/contrib/terminal/browser/terminalInstanceService.ts:48]: RemoteTerminalBackendContribution is registered to be created on restored phase but instantiation only occurs when idle.

**Comment by jeanp413** [comment]
ping @Tyriar

**Review by Tyriar — APPROVED** [review]
LGTM! Thanks for looking into this, very long standing and annoying issue ❤️

**Comment by Tyriar** [comment]
FYI @jkeech, this was a elusive Codespaces bug that had some partial fixes previously.

**Review by bpasero — APPROVED** [review]

**Review by alexr00 — APPROVED** [review]

**Pull request merged**